Bad links exporter

elenatanasoiu opened the pull request:
For [Trello card](https://trello.com/c/SYgBUl4q/258-develop-an-upload-to-ga-of-link-checker-data)

## WHAT

This PR makes a list of non-200 status links (aka "bad links") available to GA by generating a CSV file with these links and then uploading it to Google Analytics as a Custom Import. We're using the Google Analytics Management API to do this. 

You will need a `local_env.yaml` file for testing the upload locally. Be advised that the data being uploaded to Google Analytics will affect the data that already exists there. Please only upload empty data when testing, like so: 

```
data = "ga:dimension36,ga:dimension37
'',''
"
```

The app assumes five environment variables are made available through `govuk-secrets`: 
- `GOOGLE_CLIENT_EMAIL` (this is extracted from the JSON file containing the credentials for your service account, in step 1)
- `GOOGLE_PRIVATE_KEY` (this is extracted from the JSON file containing the credentials for your service account, in step 1)
- `GOOGLE_EXPORT_ACCOUNT_ID` (this is obtained from the GA account domain you are trying to upload to. We use the account id for `www.gov.uk`)
- `GOOGLE_EXPORT_TRACKER_ID` (this is the web property UA string associated with the upload. Ex: `UA-XXXXXXXXX-X`)
- `GOOGLE_EXPORT_CUSTOM_DATA_IMPORT_SOURCE_ID` (this is obtained from the custom data import setup in GA. Ex: `abcdeFGHijk12_AB12cd`)

More info in `docs/google_analytics_custom_import.md`. An update to the docs was added in a later PR: https://github.com/alphagov/local-links-manager/pull/134

That update described how you can only use one service account at a time, so there's no point in creating multiple ones. A workaround is recommended in the updated version of the docs. 

## WHY
 
The final goal is to make data from local-links-manager to non-specialists, aka in our case people who are using GA. The data can then be scheduled to be uploaded once a week. 

## CAVEATS

During the development of this PR, we wanted to take the CSV data and join it with existing hit data that we have on links. The links would be stored in the "event action" field. 
We discovered two things:
1. You can't do joins on that particular field, only on data in custom dimensions. The GA interface is designed in such a way that you can only pick custom dimensions when creating joins. 
2. It's not possible to create a filter in GA to duplicate the "event_action" data into a custom dimension. The filter will be applied before the data from the upload is loaded, so the filter won't know which data to join with as it won't be present when it uploads. 

The workaround for this will be, moving forward, to report the url in a custom dimension when we're sending hit data. 

## Further reading 

- Google Apps Service accounts: https://developers.google.com/api-client-library/ruby/auth/service-accounts
- Authorization levels for the GA Management API: https://developers.google.com/analytics/devguides/config/mgmt/v3/authorization
- Setting up a custom data import: https://developers.google.com/analytics/devguides/config/mgmt/v3/data-import
- Content data import example: https://support.google.com/analytics/answer/3271771
- Google API Client Library (Ruby): https://developers.google.com/api-client-library/ruby/start/get_started

## GA Config

The custom data import is set up as under the name "External Link status" in GA.

elenatanasoiu committed:
Rename AnalyticsService to AnalyticsImportService

This will avoid a clash with the other analytics service that will be used for exporting a list of bad links to GA.
This second service will be called the AnalyticsExportService.

elenatanasoiu committed:
Adds version of bad links export with GA headers

This will be used to create a file that will be exported to GA using the Google Analytics API.
The file will contain a list of bad links with the headers expected by Google Analytics instead of the usual ones.

elenatanasoiu committed:
Add GA export service for bad links

This will export a list of bad links that is obtained from the BadLinksUrlandStatusExporter.
The list is in CSV format and contains the page url and status, and indicates why a link is currently failing.
In this case "bad" means the link has a non-200 status.

In order to be able to upload files, a service account was created called "BadLinksExporter". This account has permission to upload to GA,
which is granted through a json file that you can download from the service account. The information you will need to extract from this file
will be the `client_email` and the `private_key`. You will then add these in a yml file which is described in detail below.

Running the GA code locally can be done by creating a file at `config/local_env.yml` which should contain:
- a `GOOGLE_CLIENT_EMAIL` (this is extracted from the JSON file containing the credentials for your service account)
- a `GOOGLE_PRIVATE_KEY` (this is extracted from the JSON file containing the credentials for your service account)
- a `GOOGLE_EXPORT_ACCOUNT_ID` (this is obtained from the GA account domain you are trying to upload to. We use the account id for `www.gov.uk`)
- a `GOOGLE_EXPORT_TRACKER_ID` (this is the web property UA string associated with the upload. Ex: `UA-XXXXXXXXX-X`)
- a `GOOGLE_EXPORT_CUSTOM_DATA_IMPORT_SOURCE_ID` (this is obtained from the custom data import setup in GA. Ex: `abcdeFGHijk12_AB12cd`)
These values are set as environment variables.

Once authentication has been obtained, the end result will be an uploaded CSV file to GA via the Analytics API.

elenatanasoiu committed:
Adds an exporter to use the analytics service

The AnalyticsExporter will manage errors when calling the service, as well as obtaining the CSV file from the BadLinksUrlAndStatusExporter.

At the same time, the export can then be called by saying `AnalyticsExporter.export` instead of concerning ourselves with getting authentication by building the service first,
and then being able to actually do the upload.

elenatanasoiu committed:
Adds rake task for bad links exporter

Running `export:google_analytics:bad_links` will upload a CSV file to GA with a list of bad links.

h-lame commented on lib/local-links-manager/export/bad_links_url_and_status_exporter.rb:
```diff
@@ -22,6 +23,15 @@ def self.bad_links_url_and_status_csv
         end
       end
+
+      def self.ga_bad_links_url_and_status_csv
```
Given we've tied the these methods together in the specs, could we tie them together in implementation too?  Perhaps a private bad_links_csv method that takes a set of headings?  I don't know how strongly I feel about doing this though - I'm not sure if *they* are the same report with different headings, or just co-incidentally are the same report right now.

h-lame commented on lib/google_analytics/analytics_export_service.rb:
```diff
@@ -0,0 +1,31 @@
+
+    def build
+      @service = AnalyticsService.new
+      @service.authorization ||= ServiceAccountCredentials.make_creds(
```
We might be able to avoid storing the `.json` file.  We were able to do so in feedback (see: https://github.com/alphagov/feedback/commit/ebe2a359be75ebf339dafbb405cdbc039cf61218), maybe we can do the same here - depending on which keys and values from the json file we actually need.

This would have the advantage of being able to manage the creds in puppet more easily.

h-lame commented on lib/google_analytics/analytics_export_service.rb:
```diff
@@ -0,0 +1,31 @@
+    end
+
+    def export_bad_links(data)
+      File.open('bad_links.csv', 'w') { |file| file.write(data) }
```
Maybe we could use a `TempFile` here to avoid colliding if two or more of these run at the same time?

h-lame commented on spec/factories/upload_response_factory.rb:
```diff
@@ -0,0 +1,13 @@
+module GoogleAnalytics
```
Do we need this?  Could it just be an rspec `double`?

h-lame commented on lib/tasks/export/analytics_bad_links_exporter.rake:
```diff
@@ -0,0 +1,23 @@
+      service_desc = "Export bad links status to Google Analytics"
+      begin
+        Rails.logger.info("Starting link exporter")
+        Services.icinga_check(service_desc, true, "Starting GA bad link exporter to")
```
I feel like the description here is truncated.  "Starting GA bad link exporter to" what?

h-lame reviewed (changes requested):
Looks good.  I have a few short comments about the implementation that may or may not be useful.  The major thing is about the JSON file vs. using env vars to achieve the same thing.

elenatanasoiu committed:
Update authentication to use update googleauth gem

This is now using the fixed version of the `googleauth` gem in order to authenticate to the Google Management API.
We're updating our call to get authorization to actually use the `get_application_default` method from the Google API instead of digging into the API to use the `make_creds` method, which was a solution we've also used in the past in here: https://github.com/alphagov/feedback/commit/ebe2a359be75ebf339dafbb405cdbc039cf61218

elenatanasoiu commented:
I've added a new commit for updating the `googleauth` gem and then an additional commit to make use of the fixed authentication.

h-lame commented on lib/local-links-manager/export/bad_links_url_and_status_exporter.rb:
```diff
@@ -14,9 +15,9 @@ def self.local_authority_bad_homepage_url_and_status_csv
       end
 
-      def self.bad_links_url_and_status_csv
+      def self.bad_links_url_and_status_csv(with_ga_headings = false)
```
The tests suggest you want this as a keyword arg (you're doing `exporter.bad_links_url_and_status_csv(with_ga_headings: true)`) but this is a normal positional arg.  The reason is works is that the arg in the test is turned into a hash, so `with_ga_headings` ends up being `{with_ga_headings: true}` which _is_ a truthy value, so we get the right output, but for the wrong reasons.

elenatanasoiu replied:
Fixed.

h-lame commented on lib/google_analytics/analytics_export_service.rb:
```diff
@@ -0,0 +1,36 @@
+
+    def export_bad_links(data)
+      begin
+        file = Tempfile.new('bad_links.csv')
```
I *think* if you use the block form of [`Tempfile.create`](https://ruby-doc.org/stdlib-2.3.0/libdoc/tempfile/rdoc/Tempfile.html#method-c-create) it cleans everything up for you, even if there is an exception.  Worth checking.

elenatanasoiu replied:
Yes it does!

h-lame commented on lib/tasks/export/analytics_bad_links_exporter.rake:
```diff
@@ -7,7 +7,7 @@ namespace :export do
       begin
         Rails.logger.info("Starting link exporter")
-        Services.icinga_check(service_desc, true, "Exporting bad links list to Google Analytics")
+        Services.icinga_check(service_desc, true, "Exporting list of bad links to Google Analytics")
```
Nitpick, but this change could be squashed down into edbd040b364e182878eb3976e56c71142c2d6bdf instead of living in this commit.

h-lame commented on docs/google_analytics_custom_import.md:
```diff
@@ -0,0 +1,23 @@
+# Authorisation for Google Analytics custom import
```
Nice to see some docs for this :+1:

h-lame commented on docs/google_analytics_custom_import.md:
```diff
@@ -0,0 +1,23 @@
+2. Extract the `client_email` value from the JSON key and make it available to the app in the "GOOGLE_CLIENT_EMAIL" environment variable.
+3. Extract the `private_key` value from the JSON key and make it available to the app in the "GOOGLE_PRIVATE_KEY" environment variable. 
+4. Make sure your local_env.yml file contains an entry for: 
+`GOOGLE_ACCOUNT_TYPE: 'service_account'`
```
`local_env.yml` is only used in dev mode, but from reading the google lib docs it looks like we'll need the `GOOGLE_ACCOUNT_TYPE` env var set properly for all envs.  What's our plan for production, and should we mention it here?  Maybe we can just assume it's always a 'service_account' and explicitly set the env var to that in code?

elenatanasoiu replied:
I've set the ENV variable for `GOOGLE_ACOUNT_TYPE` in the code, so it will always be set to `service_account` before we attempt authentication. 

I'm going to add the other five environment variables to `govuk_secrets`:
`GOOGLE_CLIENT_EMAIL`
`GOOGLE_PRIVATE_KEY`
`GOOGLE_EXPORT_ACCOUNT_ID`
`GOOGLE_EXPORT_CUSTOM_DATA_IMPORT_SOURCE_ID`
`GOOGLE_EXPORT_TRACKER_ID`

h-lame reviewed (changes requested):
Previous changes I found are all addressed, but I found more.  Sorry!

h-lame commented on docs/google_analytics_custom_import.md:
```diff
@@ -18,6 +15,32 @@
+GOOGLE_EXPORT_ACCOUNT_ID: '1234'
+GOOGLE_EXPORT_CUSTOM_DATA_IMPORT_SOURCE_ID: 'UA-12345-2'
+GOOGLE_EXPORT_TRACKER_ID: 'abcdef'
+
```
I think it might be more "correct" to combine 5 & 6 here.  Something like:

> 5.  The app assumes the following environment variables are set and made available to it:
>
>  [LIST OF ENV VARS + DESCRIPTIONS AS IS]
> 
> In development mode you can add these to your `local_env.yml` file and they will be turned into environment variables for you.  In deployed environments we manage these through govuk-puppet and govuk-secrets.

h-lame commented on lib/google_analytics/google_analytics_export_credentials.rb:
```diff
@@ -0,0 +1,15 @@
+
+    def self.all_configuration_in_env?
+      %w(GOOGLE_PRIVATE_KEY GOOGLE_CLIENT_EMAIL).all? { |env_var| ENV[env_var].present? }
+    end
```
Maybe we should add `GOOGLE_ACCOUNT_TYPE` to these lists?

elenatanasoiu replied:
I've moved the setting of this var inside `GoogleAnalyticsExportCredentials`, so now we always have it set.

h-lame replied:
:+1:

h-lame commented on config/secrets.yml:
```diff
@@ -13,13 +13,28 @@
+  GOOGLE_PRIVATE_KEY: 'abcdefg'
+  GOOGLE_EXPORT_ACCOUNT_ID: '12345'
+  GOOGLE_EXPORT_CUSTOM_DATA_IMPORT_SOURCE_ID: 'UA-XXXXX-X'
+  GOOGLE_EXPORT_TRACKER_ID: 'abcdefg'
```
Don't think we want these entries in secrets.yml now we're saying we will use env vars.

h-lame commented on lib/google_analytics/analytics_export_service.rb:
```diff
@@ -18,11 +18,12 @@ def build
+          response = service.upload_data(ENV['GOOGLE_EXPORT_ACCOUNT_ID'],
+                                         ENV['GOOGLE_EXPORT_CUSTOM_DATA_IMPORT_SOURCE_ID'],
+                                         ENV['GOOGLE_EXPORT_TRACKER_ID'],
                                          fields: "accountId,customDataSourceId,errors,id,kind,status,uploadTime",
```
This is another git hygiene nitpick, but this change seems unrelated to the rest of this commit, and could happily live in the commit where you introduce this code: e6362afe5100e286606cf6455eea2cf4b51579b9.

elenatanasoiu committed:
Update googleauth gem to fixed auth version

This will update the googleauth gem to version 0.5.3 where the usage of
`Google::Auth.get_application_default(scopes)`
has been fixed. In which case we don't have to go digging in the google API to generate valid credentials like we've done here: https://github.com/alphagov/feedback/blob/ebe2a359be75ebf339dafbb405cdbc

In order for the `google-api-client` gem to work with the new version of `googleauth`, we've had to bump the version up from 0.10.0 to 0.17.0 (which is currently the latest).

This in turn has led to a cascade of other gems having to be **conservatively** updated in order to work with either `googleauth` or `google-api-client`.

h-lame reviewed (approved):
Thanks for dealing with all my pedantry.  Looks good!
